bluetooth/src/commonMain/kotlin/scanner/ScanningState.kt

```diff
@@ -261,63 +338,139 @@ sealed class ScanningState : KalugaState {
 
-            fun permit(enabled: Boolean): suspend () -> ScanningState = {
+            override fun permit(enabled: Boolean): suspend () -> ScanningState = {
                 if (enabled) Enabled.Idle(nothingDiscovered, scanner)
                 else Disabled(scanner)
             }
         }
     }
 
-    object NoHardware : ScanningState()
+    object NoHardware : ScanningStateImpl(), ScanningState.NoHardware
 }
 
 typealias ScanningStateFlowRepo = StateRepo<ScanningState, MutableStateFlow<ScanningState>>
 
-class ScanningStateRepo(
-    permissions: Permissions,
-    connectionSettings: ConnectionSettings,
-    autoRequestPermission: Boolean,
-    autoEnableBluetooth: Boolean,
-    builder: BaseScanner.Builder,
+abstract class BaseScanningStateRepo(
+    createNotInitializedState: () -> ScanningState.NotInitialized,
+    createInitializingState: ColdStateFlowRepo<ScanningState>.(ScanningState.Inactive) -> suspend () -> ScanningState,
+    createDeinitializingState: ColdStateFlowRepo<ScanningState>.(ScanningState.Active) -> suspend () -> ScanningState.Deinitialized,
     coroutineContext: CoroutineContext = Dispatchers.Main.immediate
 ) : ColdStateFlowRepo<ScanningState>(
     coroutineContext = coroutineContext,
     initChangeStateWithRepo = { state, repo ->
         when (state) {
-            is ScanningState.NotInitialized -> {
-                val scanner = builder.create(
-                    permissions,
-                    connectionSettings,
-                    autoRequestPermission,
-                    autoEnableBluetooth,
-                    scanningStateRepo = repo as ScanningStateRepo
-                )
-                state.startInitializing(scanner)
-            }
-            is ScanningState.Deinitialized -> {
-                state.reinitialize
+            is ScanningState.Inactive -> {
+                repo.createInitializingState(state)
             }
             is ScanningState.Active, is ScanningState.NoHardware -> state.remain()
         }
     },
-    deinitChangeStateWithRepo = { state, _ ->
+    deinitChangeStateWithRepo = { state, repo ->
         when (state) {
-            is ScanningState.Active -> state.deinitialize
+            is ScanningState.Active -> repo.createDeinitializingState(state)
             is ScanningState.Inactive, is ScanningState.NoHardware -> state.remain()
         }
     },
-    firstState = {
-        ScanningState.NotInitialized()
+    firstState = createNotInitializedState
+)
+
+open class ScanningStateImplRepo(
+    createScanner: () -> Scanner,
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate,
+    private val contextCreator: CoroutineContext.(String) -> CoroutineContext = { this + singleThreadDispatcher(it) },
+) : BaseScanningStateRepo(
+    createNotInitializedState = { ScanningStateImpl.NotInitialized },
+    createInitializingState = { state ->
+        when (val stateImpl = state as ScanningStateImpl.Inactive) {
```

Comment:
Better name is `inactiveState`

Reply:
Ah, well to be honest the whole `as` is bad. I will remove it (so this will be fixed), but for that we need to move on to the next P.